index.bs

```diff
@@ -1749,38 +1750,35 @@ The [=remote end event trigger=] is:
 <div algorithm="remote end event trigger for browsingContext.contexDestroyed">
 
 Whenever the [=a browsing context is discarded=] algorithm is invoked, run the
-following steps after all {{Document}} objects have been discarded, but before
-the [=browsing context group is removed=] (if relevant):
+following steps before running any steps in HTML:
 
-Note: The intent is that when a subtree of browsing contexts is removed, the
-<code>contextDestroyed</code> event is always emitted for a child before the
-parent.
+ 1. If the [=current session=] is null, return.
 
- 1. Let |context| be the browsing context being discarded.
+ 1. If the [=current session=]'s [=context discard=] flag is set, return.
```

Comment:
Rather than maintaining this flag and "a browsing context is discarded" at the right point in the algorithm to achieve the effect of getting called only once for the topmost browsing context being discarded, can we invent a "tree of browsing contexts discarded" hook that we wish that HTML had, and later go about adding it to HTML or letting HTML call WebDriver BiDi directly?

Reply:
We can of course; do you have a concrete example of what that should look like? Are you basically just imagining that the current discard a browsing context algorithm is split into two parts, one that's called externally and one that's called recursively for nested contexts? (maybe we could just handwave about the event happening when the algorithm is called but it isn't a recursive invocation, until the hook is ready)

Reply:
https://wicg.github.io/origin-policy/#monkeypatches is the style of handwavery I have in mind, patches that are explicitly intended to graduate into other specs.

We would probably patch https://html.spec.whatwg.org/multipage/window-object.html#a-browsing-context-is-discarded, there are many ways to do it, but the net effect should be a single call to WebDriver BiDi with the topmost browsing context about to be discarded, and no calls for any of its descendant browsing contexts.

Additional examples of monkey patching separated out quite nicely:
https://w3c.github.io/webappsec-csp/#integrations
https://w3c.github.io/webappsec-referrer-policy/#integration-with-html
https://w3c.github.io/webappsec-trusted-types/dist/spec/#integration-with-html